File: Fair/WebUI/src/types/proposals/BaseProposal.ts
import { AccountBase } from "types"
import { OperationType } from "types/OperationType"

import { ProposalOption } from "./ProposalOption"

export type BaseProposal = {
  id: string

  operation: OperationType

  optionsVotesCount: number[]
  neitherCount: number
  anyCount: number
  banCount: number
  banishCount: number

  creationTime: number

  title: string
  text: string

  options: ProposalOption[]

  by: AccountBase
  multipleOptions: boolean

  hoursLeft: number
}

File: Fair/cs/WebServer/Models/Proposals/BaseProposal.cs
namespace Uccs.Fair;

public abstract class BaseProposal(Proposal proposal)
{
	public string Id { get; } = proposal.Id.ToString();

	public string Title { get; } = proposal.Title;
	public string Text { get; } = proposal.Text;

	public int CreationTime { get; } = proposal.CreationTime.Days;

	public IEnumerable<ProposalOptionModel> Options { get; set; } = null!;

	public IEnumerable<int> OptionVotesCount { get; set; } = proposal.Options.Select(x => x.Yes.Count());
	public int NeitherCount { get; set; } = proposal.Neither.Count();
	public int AbstainedCount { get; set; } = proposal.Any.Count();
	public int BanCount { get; set; } = proposal.Ban.Count();
	public int BanishCount { get; set; } = proposal.Banish.Count();
}
